man/jabberd.8.in
.TH jabberd 8 "26 Jun 2004" "@VERSION@" "jabberd14 project"
.SH NAME
jabberd \- jabberd daemon
.SH SYNOPSIS
.B jabberd
.I [-B] [-c config] [-D mask] [-H dir] [-v] [-V] [-Z zone(s)]
.SH DESCRIPTION
.BR jabberd
is the core of the jabber server containing the routing functionality that
is extended by specific logic, that is loaded from shared objects specified
in the configuration file.
.SH OPTIONS
.TP
.B \-B
Background the jabberd process. (This option is disabled in debugging mode.)
.TP
.B \-c
Specify an alternate configuration file. Default is to use
@sysconfdir@/jabber.xml
.TP
.B \-D <mask>
Output debugging information while running. This option disabled
backgrounding of the jabberd process. The debugging output is either
written to syslog or to the standard error output (configured in
jabber.xml). The given mask is ORed with the mask given in the
configuration file and each bit in the result enables a category of
debug logging messages. You can further limit the information that
is printed by using the -Z option.
.TP
.B \-H
Resolve paths relative to this directory.
.TP
.B \-v \-V
Print jabberd version information.
.TP
.B \-Z <zones>
Limit the debugging output to a given set of zones in the source code.
Comma separated list of zones. Zones are the file names of the source
files without the extension. (E.g. debugging statements in 'jabberd.c'
correspond to the zone 'jabberd'.)
.SH SEE ALSO
.BR jabber.xml (5)
.SH AUTHOR
Jabber Software Foundation


NEWS
Version ?.?.? - not released yet

Changes in administration of the server
    * Jabberd14 now understands the protocol between c2s and session manager
      components of jabberd2. This allows development and usage of
      c2s components for both server implementations.
    * It is possible to block account names from being registered and
      to enforce minimum and maximum lengths of the username on
      registration of new accounts
    * New base handler base_dir that can periodically check a directory for
      *.stanza files. These files are read, parsed, the content is
      processed as a stanza, and the file is deleted afterwards. This can
      be useful to inject messages and other stanzas to the server.
      Stanzas sent to the handler are written as files using the *.out
      extension. The directories for reading and writing are configured
      using the <in/> and <out/> elements inside the <dir/> element.
    * Settings in the dialback component configuration for disabling
      xmpp streams or disabling tls now also affect subdomains of the
      configured host. It is now also possible to require s2s connections
      to some or all other hosts to be encrypted.
    * Jabberd14 now supports SASL EXTERNAL authentication using TLS
      certificates on s2s links.
    * Jabberd14 now supports using GNU TLS as an alternative to OpenSSL.
    * xdb_sql can be configured to execute an SQL query just after a
      connection to the database server has been established or reestablished.
      This is useful for example if you are using MySQL 4.1+ and want to
      set your used charset ("SET NAMES utf8").
    * jabberd14 now uses libpopt for command line parsing.
    * jabberd14 now detects stale pidfiles, and ignores them.
    * jabberd14 does not cache passwords in the session manager anymore,
      you can now just update passwords in SQL and they are used instantly.
    * The session manager of jabberd14 can no be restarted without users
      sessions being dropped. This allows reconfiguring of the session manager
      while it is serving users.
    * The list of online users has to be fetched using service discovery
      instead of browsing now.
    * Removed support for the jabber:iq:admin namespace.
    * Disabled support for the jabber:iq:agent and jabber:iq:agents in the
      default configuration file.
    * Removed support for the jabber:iq:filter namespace (which had already
      been disabled in the default configuration of jabberd14 1.4.4).
    * Removed the mod_groups module.
    * jabberd14 automatically generates the list of supported features for
      a reply to a service discovery request. There is no need anymore to
      configure a disco result in the configuration file.

Changes in features visible to the user
    * Support for Flexible Offline Message Retrieval (XEP-0013)
    * FIX: Handling of presences with negative priority had bugs, e.g.
      messages were not delivered when changing from negative to
      non-negative priority.
    * jabberd14 now handles XML namespaces
    * jabberd14 now handles xml:lang attributes defined on the stream root
      element, declaring the default language of the stream. If a stanza does
      not redeclare the language, this language is passed along with the
      stanza.
    * The full stanza that requests subscription to a user is passed
      to this user, even if it has been stored offline. In addition the
      request is stamped if the time of the request, if it is delivered
      from offline storage.


Version 1.4.4 - 2005-04-16

Changes in compiling and installing jabberd
    * Jabberd14 now uses an automake/autoconf build environment, so you
      will use "make install" after compiling jabberd and after that
      you can now delete the source.
    * Jabberd14 now requires at least openssl version 0.9.7 to be build with
      support for SSL/TLS.

Changes in administration of the server
    * Jabberd14 now supports encrypted connections using the STARTTLS
      stream feature of XMPP.
    * The activation of debugging output has been changed: you can either
      activate debugging in the configuration file or in the command line
      using the -D option (that now takes an argument). You can even change
      the debugging settings at the configuration file while jabberd is
      running. After you changed them send a HUP signal to jabberd and
      it will reread them from the configuration.
    * The size of the files handled by xdb_file can be limited. This prevents
      spool files from getting to big if a user subscribes to a feed and stops
      using his account. Incoming messages will get bounced with an error if
      the spool file of this user is full.
    * The xdb_file module allows two directory layouts. The classic jabberd14
      layout with one big directory containing all files for one domain, and
      the layout adopted from wpjabber, where the files are distributed over
      256 * 256 different directories, improoving performance on filesystems,
      that are not well suited for large directories.
    * There is a small utility xdbfiletool that helps accessing xdb_file's
      new directory layout and that can convert old spools to the new
      format.
    * jabberd14 has a new xdb module for storing data in SQL databases,
      currently mysql and postgresql are supported
    * Support for legacy jabberd 1.0 server to server connections is now
      removed completely as this old protocol allowed spoofing of sender
      addresses. You should not need this protocol anymore anyway.
    * Jabberd14 will now try to establish s2s connections not only on the
      first IPv4/IPv6 address it finds in DNS but on any IPv4/IPv6 address,
      that can be found for the first matching service. (It will not fall
      back to pure AAAA or A records if there has been a SRV record for
      any configured service.)
      This will now allow a jabberd14, that has been configured with

jabberd/jabberd.cc
/* --------------------------------------------------------------------------
 *
 * License
 *
 * The contents of this file are subject to the Jabber Open Source License
 * Version 1.0 (the "JOSL").  You may not copy or use this file, in either
 * source code or executable form, except in compliance with the JOSL. You
 * may obtain a copy of the JOSL at http://www.jabber.org/ or at
 * http://www.opensource.org/.  
 *
 * Software distributed under the JOSL is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied.  See the JOSL
 * for the specific language governing rights and limitations under the
 * JOSL.
 *
 * Copyrights
 * 
 * Portions created by or assigned to Jabber.com, Inc. are 
 * Copyright (c) 1999-2002 Jabber.com, Inc.  All Rights Reserved.  Contact
 * information for Jabber.com, Inc. is available at http://www.jabber.com/.
 *
 * Portions Copyright (c) 1998-1999 Jeremie Miller.
 * 
 * Acknowledgements
 * 
 * Special thanks to the Jabber Open Source Contributors for their
 * suggestions and support of Jabber.
 * 
 * Alternatively, the contents of this file may be used under the terms of the
 * GNU General Public License Version 2 or later (the "GPL"), in which case
 * the provisions of the GPL are applicable instead of those above.  If you
 * wish to allow use of your version of this file only under the terms of the
 * GPL and not to allow others to use your version of this file under the JOSL,
 * indicate your decision by deleting the provisions above and replace them
 * with the notice and other provisions required by the GPL.  If you do not
 * delete the provisions above, a recipient may use your version of this file
 * under either the JOSL or the GPL. 
 * 
 * 
 * --------------------------------------------------------------------------*/

/**
 * @file jabberd.c
 * @brief Where all begins ... this file contains the main function, shutdown routines and signal handlers
 */

/*
 * Doesn't he wish!
 *
 * <ChatBot> jer: Do you sometimes wish you were written in perl?
 *
 */

#include <pwd.h>
#include <grp.h>

#include "jabberd.h"

#ifdef HAVE_SYSLOG
#include <syslog.h>
#endif

#include <popt.h>

xht debug__zones = NULL;		/**< the debugging zones, that are enabled (key = zone string, value = zone string) */
extern int deliver__flag;
extern xmlnode greymatter__;

extern xht instance__ids;

/*** internal functions ***/
extern "C" {
    void base_init(pool p);
    int configo(int exec);
    int configurate(char *file, xht cmd_line, int is_restart);
    void deliver_init(pool p);
    void deliver_shutdown(void);
    void heartbeat_birth(void);
    void heartbeat_death(void);
    void shutdown_callbacks(void);
}
static void _jabberd_signal(int sig);
static void _jabberd_atexit(void);
static result jabberd_signal_handler(void *arg);

/**
 * structure that holds 'global' data
 */
typedef struct {
    xht		cmd_line;	/**< hash where unused command-line options are stored to */
    pool	runtime_pool;	/**< memory pool with a livetime of the jabber server */
    int		signalflag;	/**< set to the signal value, if jabberd receives a signal */
    char*	cfgfile;	/**< configuration file the user specified at the command line (NULL for the default file) */
} jabberd_struct;

jabberd_struct jabberd = { NULL, NULL, 0, NULL };		/**< global data for the jabberd */

extern "C" void xmlnode_stat();

/**
 * the entry point to jabberd
 *
 * @param argc the number of arguments on the command line used to start jabberd
 * @param argv array of the arguments
 * @return 0 on successfull shutdown, 1 else
 */
int main (int argc, const char** argv) {
    char *c = NULL;
    char *cmd = NULL;
    char *home = NULL;
    char *zones = NULL;		/* debugging zones */
    char *host = NULL;		/* domain/hostname to run as */
    char *spool = NULL;		/* spool directory for xdb_file */
    char *import_spool = NULL;	/* spool base dir for import */
    char *do_include = NULL;	/* include files in configuration */
    float avload;
    int do_debug = 0;           /* Debug output option, default no */
    int do_background = 0;      /* Daemonize option, default no */
    int do_version = 0;		/* print version information */
    char *run_as_user = NULL;	/* user to run jabberd as */
    poptContext pCtx = NULL;
    int pReturn = 0;		/* return code of popt */

    /*
     * command line options for jabberd14
     */
    struct poptOption options[] = {
	{ "config", 'c', POPT_ARG_STRING, &(jabberd.cfgfile), 0, "configuration file to use", "path and filename"},
	{ "include", 'i', POPT_ARG_STRING, &do_include, 0, "include configuration files", "comma separated list"},
	{ "debugmask", 'd', POPT_ARG_INT, &do_debug, 0, "enable debugging (by type)", "debugging mask"},
	{ "debug", 'D', POPT_ARG_NONE, NULL, 1, "enable debugging (all types)", NULL},
	{ "zones", 'Z', POPT_ARG_STRING, &zones, 0, "debugging zones (file names without extension)", "comma separated list"},
	{ "user", 'U', POPT_ARG_STRING, &run_as_user, 0, "run " PACKAGE " as another user", "user to run as"},
	{ "home", 'H', POPT_ARG_STRING, &home, 0, "what to use as home directory", "directory path"},
	{ "define", 'x', POPT_ARG_STRING, NULL, 2, "define a replacement string for configuration", "key:value"},
	{ "background", 'B', POPT_ARG_NONE, &do_background, 0, "background the server process", NULL},
	{ "host", 'h', POPT_ARG_STRING, &host, 0, "hostname that should be served by " PACKAGE, "domain (FQDN)"},
	{ "spooldir", 's', POPT_ARG_STRING, &spool, 0, "directory where to place the file spool of xdb_file", "directory path"},
	{ "import", 'I', POPT_ARG_STRING, &import_spool, 0, "import data to the server from a filespool", "basedir of file-spool"},
	{ "version", 'V', POPT_ARG_NONE, &do_version, 0, "print server version", NULL},
	{ NULL, 'v', POPT_ARG_NONE|POPT_ARGFLAG_DOC_HIDDEN, &do_version, 0, "print server version", NULL},
	POPT_AUTOHELP
	POPT_TABLEEND
    };

    /* create hash for command line options */
    jabberd.cmd_line = xhash_new(11);
    
    /* parse command line options */
    pCtx = poptGetContext(NULL, argc, argv, options, 0);
    while ((pReturn = poptGetNextOpt(pCtx)) >= 0) {
	switch (pReturn) {
	    case 1:
		do_debug = -1;
		break;
	    case 2:
		cmd = pstrdup(jabberd.cmd_line->p, poptGetOptArg(pCtx));
		c = strchr(cmd, ':');
		if (c == NULL) {
		    fprintf(stderr, "Invalid definition for config file replacement: %s\nNeeds to be of key:value\n", cmd);
		    return 1;
		}
		c[0] = 0;
		c++;
		xhash_put(jabberd.cmd_line, cmd, c);
		cmd = c = NULL;
		break;
	}
    }

    /* error? */
    if (pReturn < -1) {
	fprintf(stderr, "%s: %s\n", poptBadOption(pCtx, POPT_BADOPTION_NOALIAS), poptStrerror(pReturn));
	return 1;
    }

    /* anything left? */
    if (poptPeekArg(pCtx) != NULL) {
	fprintf(stderr, "invalid argument: %s\n", poptGetArg(pCtx));
	return 1;
    }

    /* printing version information desired? */
    if (do_version != 0) {
	printf("%s version %s\n", PACKAGE, VERSION);
	printf("\nThe following optional features have been enabled:\n");
#ifdef WITH_IPV6
	printf("- support for IPv6.\n");
#endif
	printf("- support for TLS");
	printf(" using GNU TLS");
	printf("\n");
#ifdef HAVE_MYSQL
	printf("- support for MySQL\n");
#endif
#ifdef HAVE_POSTGRESQL
	printf("- support for PostgreSQL\n");
#endif
#ifdef HAVE_SYSLOG
	printf("- logging to syslog\n");
#endif
	printf("\nDefault config file is: %s\n", CONFIG_DIR "/jabber.xml");
	printf("Locales are in: %s\n", LOCALEDIR);
	printf("\nFor more information please visit http://jabberd.org/\n");
	printf("If you need support, check out http://jabberd.org/gettingSupport\n");
	printf("\nNOTICE: With the next release of this software, this package\n");
	printf("        will get renamed to 'xmppd'. (http://xmppd.org/)\n");
	return 0;
    }

    /* generate a memory pool that is available for the whole livetime of jabberd */
    jabberd.runtime_pool = pool_new();

    /* register this handler to remove our pidfile at exit */
    atexit(_jabberd_atexit);

    /* putting h and s to the cmd_line hash */
    if (host != NULL) {
	xhash_put(jabberd.cmd_line, "h", host);
    }
    if (spool != NULL) {
	xhash_put(jabberd.cmd_line, "s", spool);
    }
    if (do_include != NULL) {
	xhash_put(jabberd.cmd_line, "i", do_include);
    }
    if (import_spool != NULL) {
	xhash_put(jabberd.cmd_line, "I", import_spool);
    }

    /* the special -Z flag provides a list of zones to filter debug output for, flagged w/ a simple hash */
    if (zones != NULL) {
	debug__zones = xhash_new(11);
	cmd = pstrdup(debug__zones->p, zones);
        while(cmd != NULL) {
            c = strchr(cmd, ',');
            if (c != NULL) {
                *c = '\0';
                c++;
            }
            xhash_put(debug__zones, cmd, cmd);
            cmd = c;
        }
    } else {
        debug__zones = NULL;
    }

    if (do_debug && do_background) {
	printf(PACKAGE " will not background with debugging enabled.\n");
	do_background=0;
    }

#ifdef HAVE_SYSLOG
    openlog(PACKAGE, LOG_PID, LOG_DAEMON);
#endif

    /* set to debug mode if we have it */
    set_cmdline_debug_flag(do_debug);

    /* Switch to the specified user */
    if (run_as_user != NULL) {
        struct passwd* user = NULL;

        user = getpwnam(run_as_user);
        if (user == NULL) {
            fprintf(stderr, "Unable to lookup user %s.\n", cmd);
            exit(1);
        }
        
        if (setgid(user->pw_gid) < 0) {
            fprintf(stderr, "Unable to set group permissions.\n");
            exit(1);
        }
        if (setuid(user->pw_uid) < 0) {
            fprintf(stderr, "Unable to set user permissions.\n");
            exit(1);
        }
    }

    /* change the current working directory so everything is "local" */
    if(home != NULL && chdir(home))
        fprintf(stderr, "Unable to access home folder %s: %s\n", home, strerror(errno));

    /* background ourselves if we have been flagged to do so */
    if (do_background != 0) {
        if (fork() != 0) {
            exit(0);
        }
    }

    /* load the config passing the file if it was manually set */
    if(configurate(jabberd.cfgfile, jabberd.cmd_line, 0))
        exit(1);

    /* EPIPE is easier to handle than a signal */
    signal(SIGPIPE, SIG_IGN);

    /* handle signals */
    signal(SIGHUP,_jabberd_signal);
    signal(SIGINT,_jabberd_signal);
    signal(SIGTERM,_jabberd_signal);

    /* init pth */
    pth_init();

#ifdef LIBIDN
    /* init the stringprep caches for jid manipulation */
    jid_init_cache();
#endif

    /* fire em up baby! */
    heartbeat_birth();

    /* register a function that regularily checks the signal flag */
    register_beat(1, jabberd_signal_handler, &jabberd);

    /* init MIO */
    mio_init();

    base_init(jabberd.runtime_pool);
    deliver_init(jabberd.runtime_pool);

    /* everything should be registered for the config pass, validate */
    deliver__flag = 0; /* pause deliver() while starting up */
    if (configo(0))
        exit(1);

    log_notice(NULL, "initializing server");

    /* karma granted, rock on */
    if(configo(1))
        exit(1);

    /* begin delivery of queued msgs */
    deliver__flag=1;
    deliver(NULL,NULL);

    /* was there a request to import an existing filespool? */
    if (import_spool != NULL) {
    }

    while (1) {
        pth_ctrl(PTH_CTRL_GETAVLOAD, &avload);
        log_debug2(ZONE, LOGT_STATUS, "main load check of %.2f with %ld total threads", avload, pth_ctrl(PTH_CTRL_GETTHREADS));
#ifdef POOL_DEBUG
	pool_stat(0);
	xmlnode_stat();
#endif
#ifdef LIBIDN
	jid_clean_cache();
#endif
        pth_sleep(60);
    };

    /* we never get here */
    return 0;
}

/**
 * jabberd signal handler
 *
 * puts the received signal in the ::jabberd__signalflag variable and returns
 */
static void _jabberd_signal(int sig) {
    log_debug2(ZONE, LOGT_EVENT, "received signal %d", sig);
    jabberd.signalflag = sig;
}

/**
 * reload the configuration file
 *
 * this gets called if jabberd receives a SIGHUP signal
 *
 * The configuration file is reloaded, but not much is done with the reloaded configuration file yet :(
 */
static void _jabberd_restart(jabberd_struct *j) {
    xmlnode temp_greymatter;

    log_notice(NULL, "reloading configuration");

    /* keep greymatter around till we are sure the reload is OK */
    temp_greymatter = greymatter__;

    log_debug2(ZONE, LOGT_CONFIG, "Loading new config file");

    /* try to load the config file */
    if(configurate(j->cfgfile, j->cmd_line, 1))
    { /* failed to load.. restore the greymatter */
        log_debug2(ZONE, LOGT_CONFIG, "Failed to load new config, resetting greymatter");
        log_alert(ZONE, "Failed to reload config!  Resetting internal config -- please check your configuration!");
        greymatter__ = temp_greymatter;
        return;
    }

    /* free old greymatter (NOTE: shea, right! many internal tables/callbacs/etc have old config pointers :)
    if(temp_greymatter != NULL)
        xmlnode_free(temp_greymatter); */

    /* XXX do more smarts on new config */

    log_debug2(ZONE, LOGT_CONFIG, "reload process complete");

}

/**
 * shutdown the jabberd process
 *
 * this gets called if jabberd receives any non-SIGHUP signal
 */
static void _jabberd_shutdown(void) {
    log_notice(NULL,"shutting down server");

    /* pause deliver() this sucks, cuase we lose shutdown messages */
    deliver__flag = 0;
    shutdown_callbacks();

    /* one last chance for threads to finish shutting down */
    pth_sleep(1);

    /* stop MIO and heartbeats */
    /* XXX disabled for jabberd 1.4.4 release, we get crashes with it
     * on shutdown. Care for it later
     */
    /* mio_stop(); */
    heartbeat_death();

    /* kill any leftover threads */
    pth_kill();

    /* exit jabberd, _jabberd_atexit() will be called */
    exit(0);
}

/**
 * do jobs, that have to be done even on fatal server shutdowns (if the exit() function is called)
 *
 * remove the pid file of this process, done in an atexit function
 * because there are multiple occurences of exit() where
 * _jabberd_shutdown is not called
 */
static void _jabberd_atexit(void) {
    xmlnode pidfile = NULL;
    char *pidpath = NULL;
    xht namespaces = NULL;

    /* Get rid of our pid file */
    namespaces = xhash_new(3);
    xhash_put(namespaces, "", const_cast<void*>(static_cast<const void*>(NS_JABBERD_CONFIGFILE)));
    pidfile = xmlnode_get_list_item(xmlnode_get_tags(greymatter__, "pidfile", namespaces), 0);
    xhash_free(namespaces);
    if (pidfile != NULL) {
        pidpath = xmlnode_get_data(pidfile);
        if(pidpath != NULL)
            unlink(pidpath);
    }
    xmlnode_free(greymatter__);

    /* free delivery hashes */
    deliver_shutdown();

#ifdef LIBIDN
    /* free stringprep caches */
    jid_stop_caching();
#endif

    /* free instances hash table */
    if (instance__ids != NULL)
	xhash_free(instance__ids);

    /* free command line options */
    if (jabberd.cmd_line != NULL)
	xhash_free(jabberd.cmd_line);

    /* free remaining global memory pools */
    if (jabberd.runtime_pool != NULL)
	pool_free(jabberd.runtime_pool);
    
#ifdef POOL_DEBUG
    /* print final pool statistics ... what we missed to free */
    pool_stat(1);
#endif
}

/**
 * check if a signal has been received, and process
 *
 * @param arg pointer to an int, that contains the last received signal
 */
static result jabberd_signal_handler(void *arg) {
    jabberd_struct *j = (jabberd_struct *)arg;

    /* no flag value location is given ... giving checks up */
    if (j == NULL)
	return r_UNREG;

    /* action based on the content of the signalflag */
    switch (j->signalflag) {
	case 0:		/* no signal received */
	    break;
	case SIGHUP:
	    _jabberd_restart(j);
	    j->signalflag = 0;
	    break;
	default:
	    _jabberd_shutdown();
    }

    return r_DONE;
}
